add default shellHook

zimbatm opened the pull request:
This hook is a destructive as it replaces the node_modules folder in
the current directory. Fortunately the node_modules is not critical and
can be re-created using `yarn`.

zimbatm commented:
Lets' discuss if this is the right default behaviour before committing to this design.

moretea commented:
@zimbatm what about adding an `enableYarnShellHook` input, which will enable your `shellHook`? I guess the most important thing here, is that people should opt-in into this, per-project. What do you think?

frontsideair commented:
My original solution to this created a symlink of `node_modules` to cwd and would cleanup on exit. It looked like this:

```
ln -sf ${dependencies}/node_modules .
tram "rm node_modules" 0
```

zimbatm commented:
@moretea we could also make the script available so that the user could do `shellHook = yarn2nix.linkNodeModulesHook`.

@frontsideair I think that there was a reason why we don't symlink the node_modules folder directly, not remembering why right now.

moretea commented:
> @moretea we could also make the script available so that the user could do shellHook = yarn2nix.linkNodeModulesHook.

Yes, that is a nicer way of doing it. If you'd create a commit for that, I'll merge it immediately :)

frontsideair commented:
@zimbatm creating a link worked for me in the past both for build and shell, I wonder what was the problem you encountered.

zimbatm committed:
add default shellHook

allows the user to do something like this:

    with yarn2nix;
    mkYarnPackage {
      src = ./.;
      shellHook = linkNodeModulesHook
    }

then when invoking nix-shell, the build node_modules is symlinked into
the current folder.

zimbatm commented:
@moretea there you go!

zimbatm commented:
a bit OT but the reason we're not symlinking in the installPhase is to be able to put the current package in the same node_modules. This might not be necessary. Right now all mkYarnPackage derivations are competing in the $out/node_modules namespace on install. Maybe we can install each package to `$out/libexec/${name}` and have the node_modules under that instead.

frontsideair commented:
I have no comment on that, but the commit looks good to me and can't wait for it to get merged.

zimbatm committed:
fixup: also remove node_modules if it's a symlink

frontsideair commented:
Awesome!

zimbatm commented:
let me know how it goes :)

frontsideair commented:
Just tried, both build and shell works as intended. 👍
